src/array.jl

```diff
@@ -389,17 +389,26 @@ end
 # Methods preserving levels and more efficient than AbstractArray fallbacks
 copy(A::CategoricalArray) = deepcopy(A)
 
-function copy!(dest::CategoricalArray{T, N}, dstart::Integer,
-               src::CategoricalArray{T, N}, sstart::Integer,
-               n::Integer=length(src)-sstart+1) where {T, N}
+function copy!(dest::CategoricalArray{<:Union{T, Null}, N},
+               dstart::Integer,
+               src::S,
+               sstart::Integer,
+               n::Integer=length(src)-sstart+1) where
+    {T, N, S <: Union{CategoricalArray{<:Union{T ,Null}, N},
+                      SubArray{<:Any, <:Any, <:CategoricalArray{<:Union{T, Null}, N}}}}
```

Comment:
IIUC, the 2nd tparam of `SubArray` is dimensionality, so it should match `N` of the destination, whereas the dimensionality of the parent should not be checked.

Reply:
How much this `copy!()` really depends on the explicit nullability? Could `<:Union{T,Null}` be simplified into `T`?
There's a use of `T` in `levels!()` below, but there you can use `leveltype(spool)`.
So in principle the declaration could be simplified to `dest::CategoricalArray{T,N}`, `src::Union{CategoricalArray{S, N}, SubArray{<:Any, N, <:CategoricalArray{S}}`, where either `T === Union{S, Null}` or `S === Union{T, Null}`, but it should be ok to do this check in the beginning of the function body.

Reply:
Regarding nullability, I think I had left a comment in the original PR. The problem is that you're allowed to copy from a nullable array into a non-nullable array. As for what's the best way of ensuring consistency, I'd say the most compact solution is the best one.

Actually it looks like the current method is too restrictive, as we should also preserve levels when the types of source and destination are different (e.g. different integer types). So we could as well make remove restrictions on the element types (but test that it works with different types, and that in case of conversion error we don't corrupt the destination).

Reply:
@cjprybol Have you tried using `CatArrSrc` here?

Reply:
So it looks like for nullability there's no other way than to do a run-time check for zeros in `src` if `dest` is non-nullable.

I'm not sure it's the most compact way, but since there's no easy way to express the requirements of `src` and `dest` category level types in the function signature, `src` could also be declared as `src::AbstractArray{<:Union{CatValue, Null}, N}`. There could be a weird case of `src` being neither `CategoricalArray` nor its wrapper, but then calls like `leveltype(src)` would fail.

Reply:
Thank you both for the helpful comments!
> @cjprybol Have you tried using CatArrSrc here?

I originally had `src::CatArrSrc` (although it was a different name before) and I couldn't get the dispatch to work correctly. The element-type `T` in the `CatArrSrc` definition wasn't being recognized as the same element-type `T` of the `dest` inside of the `copy!` function. When I copied & the `CatArrSrc` definition into the `copy!` function parameters everything worked. However, it sounds like we're all in agreement that `src` and `dest` don't need to have the same element-type `T` because we may want to copy CategoricalArrays where the eltypes of `src` and `dest` are supertypes or subtypes of one another. I may be able to use `src::CatArrSrc` again as long as I can get the dimensionality parameterization to work. Working on testing that out now